_posts/2021-06-07-spring-schedulers.md

```diff
@@ -0,0 +1,281 @@
+ @Nullable
+ volatile ScheduledFuture<?> future;
+
+ ScheduledTask(Task task) {
+   this.task = task;
+ }
+ public void cancel() {
+   ScheduledFuture<?> future = this.future;
+   if (future != null) {
+     future.cancel(true);
+   }
+ }
+}
+
+public class Task {
+ private final Runnable runnable;
+
+ public Task(Runnable runnable) {
+   Assert.notNull(runnable, "Runnable must not be null");
+   this.runnable = runnable;
+ }
+
+ public Runnable getRunnable() {
+   return this.runnable;
+ }
+}
+```
+
+Now hopefully all basics are clear, lets where all the magic happens and how these things come together to save the world
+
+```java
+ScheduledAnnotationBeanPostProcessor
+```
+ is the class, when you use `@Scheduled` annotation on any void method, this class holds the responsibility of registering that as ScheduledTask and its management.
+
+This class has a `processScheduled()` which is called when `@Scheduled` is used,
+And this class register the task, based on whatever type of flavour you have used for example **if its a `@Scheduled(cron=””)`  it will register a CronTask, if it is a `@Scheduled(fixedRate=””)` it will register a FixedRateTask.**
+
+```java
+public class CronTask extends TriggerTask {
+ private final String expression;
+
+ public CronTask(Runnable runnable, String expression) {
+   this(runnable, new CronTrigger(expression));
+ }
+
+ public CronTask(Runnable runnable, CronTrigger cronTrigger) {
+   super(runnable, cronTrigger);
+   this.expression = cronTrigger.getExpression();
+ }
+
+ public String getExpression() {
+   return this.expression;
+ }
+}
+``` 
+
+Now we won’t look at all the tasks, you can look at the processScheduled method for more detail.
+Why I wanted to show is to have your attention on **private final String expression;** 
+
+**It is final which means once we use `@Scheduled(cron=””)`, we CANNOT update it programmatically later**, if we want to change it, we will need to register a new task and stop this one.
+
+The Same goes with tasks of any other flavor, **they all are IMMUTABLE when we create it with `@Scheduled` annotation**.
+
+**Unfortunately, no annotation ssupports mutable tasks**. That means we will have to do it manually.
+
+Now to have the ability to change the scheduling at run time, we will need to use  
+```java
+ScheduledFuture schedule(Runnable task, Trigger trigger);
+```             
+Recollect we already saw this in `TaskScheduler` and that's what we will see in the next section.
+
+
+## How to schedule a task without annotation
+
+  The first question that comes to any developers mind is **Why schedule a task with annotation?**        
+
+  * **Tasks created with @Schedule annotation are immutable:** As we read in the previous section, `@Scheduled` annotation tasks are immutable, For example, we scheduled a task that pulls data at 4 AM every day, now if we want to change it on runtime, we can’t. We will need to stop the application and update the cron expression.
```

Comment:
The code example below combines two things (mutability of cron expressions and custom trigger logic). This is hard to grasp, especially since the weather forecast example is not reflected in the code example. 

Please create two sub sections "Changing a Task's Configuration at Runtime" and "Implementing Custom Trigger Logic", each with its own explanation an its own code example. That would make this text so much easier to understand.

Reply:
made suggested changes, but I am using the spring cloud config example for changing a task's configuration at runtime, is it fine or should we use a simpler example

Reply:
@thombergs